### PR TITLE
GH-95245: Store object values and dict pointers in single tagged pointer.

### PR DESCRIPTION
**Pull request by markshannon**
This is the first part of #95245.  Part two will move the weakrefs pointer into the object pre-header.

I have chosen to tag the values pointer, rather than the dict pointer, because it allows `_PyObject_GetDictPtr(PyObject *obj)` to continue to return a valid pointer.

* Issue: gh-95245

### CODE REVIEW

**Comment by markshannon** [comment]
[No change in performance](https://gist.github.com/markshannon/c79bc08277482edc9f62e4f6357d1727)

File: Objects/dictobject.c [hunk 1]
```diff
@@ -5511,105 +5540,94 @@ _PyObject_IsInstanceDictEmpty(PyObject *obj)
...
-}
-
 void
 _PyObject_FreeInstanceAttributes(PyObject *self)
 {
     PyTypeObject *tp = Py_TYPE(self);
     assert(Py_TYPE(self)->tp_flags & Py_TPFLAGS_MANAGED_DICT);
-    PyDictValues **values_ptr = _PyObject_ValuesPointer(self);
-    if (*values_ptr == NULL) {
+    PyDictOrValues dorv= *_PyObject_DictOrValuesPointer(self);
```
Comment (kumaraditya303) [inline Objects/dictobject.c]:
```suggestion
    PyDictOrValues dorv = *_PyObject_DictOrValuesPointer(self);
```

File: Objects/object.c [hunk 2]
```diff
@@ -1170,36 +1164,46 @@ _PyObject_GetMethod(PyObject *obj, PyObject *name, PyObject **method)
...
-        PyObject **dictptr = _PyObject_DictPointer(obj);
-        PyObject *dict;
-        if (dictptr != NULL && (dict = *dictptr) != NULL) {
-            Py_INCREF(dict);
-            PyObject *attr = PyDict_GetItemWithError(dict, name);
+    PyObject *dict;
+    if ((tp->tp_flags & Py_TPFLAGS_MANAGED_DICT)) {
+        PyDictOrValues* dorv_ptr = _PyObject_DictOrValuesPointer(obj);
+        if (_PyDictOrValues_IsValues(*dorv_ptr)) {
+            PyDictValues *values = _PyDictOrValues_GetValues(*_PyObject_DictOrValuesPointer(obj));
```
Comment (kumaraditya303) [inline Objects/object.c]:
```suggestion
            PyDictValues *values = _PyDictOrValues_GetValues(*dorv_ptr);
```

File: Objects/object.c [hunk 3]
```diff
@@ -1420,6 +1431,7 @@ _PyObject_GenericSetAttrWithDict(PyObject *obj, PyObject *name,
             res = PyDict_SetItem(dict, name, value);
         Py_DECREF(dict);
     }
+  error_check:
```
Comment (iritkatriel) [inline Objects/object.c:1434]: There is only one goto to this label. Could this check not be inlined? Or did you mean to jump here from line 1410?

Comment (markshannon) [reply Objects/object.c:1434]: `error_check:` is reachable by falling through from the code above as well.

Line 1410 sets the error, no need to check for it. That's why it jumps to `done`.
I'll admit the code is a bit convoluted, but I think that is indicative of the underlying complexity.

File: Objects/object.c [hunk 4]
```diff
@@ -1389,27 +1393,34 @@ _PyObject_GenericSetAttrWithDict(PyObject *obj, PyObject *name,
...
-                                tp->tp_name, name);
-                }
-                goto done;
+            dictptr = _PyObject_ComputedDictPointer(obj);
+        }
+        if (dictptr == NULL) {
+            if (descr == NULL) {
+                PyErr_Format(PyExc_AttributeError,
+                            "'%.100s' object has no attribute '%U'",
+                            tp->tp_name, name);
```
Comment (iritkatriel) [inline Objects/object.c:1413]: Why is it '%.100s' here and '%.50s' below?

Comment (markshannon) [reply Objects/object.c:1413]: I've no idea. I'm just preserving the formatting.

**Comment by iritkatriel** [comment]
Is there any documentation about this? It's pretty intricate.

**Comment by markshannon** [comment]
I was going to document this once I'd done the second part of https://github.com/python/cpython/issues/95245
Something like this: https://github.com/faster-cpython/cpython/blob/document-object-layout/Objects/object_layout.md

**Review by iritkatriel — APPROVED** [review]
I have no more comments.

File: Include/internal/pycore_object.h [hunk 5]
```diff
@@ -268,24 +268,49 @@ extern int _PyObject_StoreInstanceAttribute(PyObject *obj, PyDictValues *values,
...
+_PyDictOrValues_GetDict(PyDictOrValues dorv)
+{
+    assert(!_PyDictOrValues_IsValues(dorv));
+    return dorv.dict;
+}
+
+static inline void
+_PyDictOrValues_SetValues(PyDictOrValues *ptr, PyDictValues *values)
+{
+    ptr->values = ((char *)values) - 1;
```
Comment (matthiasgoergens) [inline Include/internal/pycore_object.h:307]: I'm curious why this one is a `-1`?  Naively, I would have expected something like (`|1`), ie setting the lowest bit.  Which would be equivalent to `+1`.

I assume there's something that goes horribly wrong with that native approach that I am missing, though?

(As far as I can tell, subtracting one should also work.  Not sure what C does with null pointers in this context, though.)

Comment (erlend-aasland) [reply Include/internal/pycore_object.h:307]: `_PyDictOrValues_SetValues` mutates the pointer; `_PyDictOrValues_GetValues` restores the pointer. Think of it as $f(x) = y$ and $f^{-1}(y) = x$. Currently, $f(x) = x - 1$ and $f^{-1}(y) = y + 1$. If we'd used `+1`, as you suggest, we'd have to undo that using `-1` iso. `+1`. AFAICS, either way is fine.

Comment (matthiasgoergens) [reply Include/internal/pycore_object.h:307]: Thanks for explaining!

I am working on a prototype with a tagged pointer elsewhere in the code, I so I was looking for exiting examples to learn from.

Thanks for confirming that either way is fine.

I was just wrecking my brain about whether there's something more going on that I was missing. (And something that I would thus be bound to get horriby wrong in my code, where I want to store either a pointer directly, or an odd integer, in the same memory location.)

**Pull request merged**